### PR TITLE
ARM SD card image expressions

### PR DESCRIPTION
**Pull request by dezgeg**
This series provides nix expressions for building bootable SD card images for ARM systems. These images come in two variants; one for the Raspberry Pi (ARMv6 / RPi 1 only) and one for ARMv7 boards supporting the multiplatform kernel (`multi_v7_defconfig`).

The resulting image can be copied to a SD card with `dd` and is directly bootable by a suitably configured U-Boot. Though depending on the board, some extra steps are required for copying U-Boot itself to the SD card.

Inside the image is a MBR partition table, with a FAT32 /boot and a normal writable EXT4 rootfs. It's possible to directly reuse the SD image's partition layout and "install" NixOS on the same SD card by replacing the default configuration.nix and running `nixos-rebuild`, and actually is the preferred way to use these images. To assist in this installation method, the boot scripts on the image automatically resize the rootfs partition to fit the SD card on the first boot. Of course, `nixos-install` is also available.

I've temporarily put up the built images at http://www.cs.helsinki.fi/u/tmtynkky/images/, but later on I'll auto-create the images from some cronjob with sane filenames, and also the U-Boots that are required. I'll put the final link into the wiki, along with some documentation; here's a draft:
# Board-specific installation steps
## Raspberry Pi

The `sd-image-raspberrypi.img` image copied to a SD card will boot without modifications.
Even though U-Boot shows up on the HDMI display, USB keyboards are not yet supported
by the RPi USB code. Also, the image won't work on a Raspberry Pi 2 yet. (It will be
supported by `sd-image-armv7l-linux.img` once RPi2 support is added to the mainline
kernel).
## Jetson TK1

The `sd-image-armv7l-linux.img` copied to a SD card or an USB stick will boot without any modifications,
but U-Boot needs to be flashed onto the board's internal eMMC memory beforehand via a microUSB cable.
The easiest way is to use https://github.com/NVIDIA/tegra-uboot-flasher-scripts, though unfortunately
that's currently not packaged in NixOS.
The only way to interact with U-Boot on this board is via the serial port.
## pcDuino3 Nano

After copying `sd-image-armv7l-linux.img` to a microSD card, an additional step is required to copy U-Boot to the card. Download `u-boot-sunxi-with-spl.bin` and run:

```
sudo dd if=u-boot-sunxi-with-spl.bin of=SDCARD bs=1024 seek=8
```
## Versatile Express (Cortex-A9) in QEMU

First, convert the downloaded `sd-image-armv7l-linux.img` to a new qcow2 formatted disk image (for example, `arm-vm.qcow2`) and resize it to the desired size (for example, `2.5G`):

```
qemu-img convert -O qcow2 sd-image-armv7l-linux.img arm-vm.qcow2
qemu-img resize vm.qcow2 2.5G
```

Then, to boot the VM, download U-Boot for the Versatile board (`uboot-vexpress_ca9x4_defconfig-2015.04_u-boot') and run:

```
qemu-system-arm -kernel uboot-vexpress_ca9x4_defconfig-2015.04_u-boot -M vexpress-a9 -serial stdio -sd arm-vm.qcow2 -m 512
```

U-Boot will be available over the emulated serial port connected to the stdio/terminal (i.e. NOT the graphical window displaying the framebuffer).
The amount of memory (`-m 512`) should be at least 512 megabytes, or you will get all sorts of funny behaviour if/when U-Boot tries to access nonexistent RAM.
Unfortunately, the maximum amount of memory is limited to 1 gigabyte.
# Configuration sample

A configuration for Raspberry Pi that reuses the SD image's partition layout:

```
{ config, pkgs, lib, ... }:
{
  boot.loader.grub.enable = false;
  boot.loader.generic-extlinux-compatible.enable = true;
  boot.kernelPackages = pkgs.linuxPackages_rpi;

  services.nixosManual.enable = false; # Manual doesn't currently evaluate on ARM

  fileSystems = {
    "/boot" = {
      device = "/dev/disk/by-label/NIXOS_BOOT";
      fsType = "vfat";
    };
    "/" = {
      device = "/dev/disk/by-label/NIXOS_SD";
      fsType = "ext4";
    };
  };
}
```

### CODE REVIEW

File: nixos/modules/installer/cd-dvd/sd-image.nix [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+    inherit pkgs;
+    inherit (config.sdImage) storePaths;
+    volumeLabel = "NIXOS_SD";
+  };
+in
+{
+  options.sdImage = {
+    storePaths = mkOption {
+      example = literalExample "[ pkgs.stdenv ]";
+      description = ''
```
Comment (bjornfor) [inline nixos/modules/installer/cd-dvd/sd-image.nix]: Maybe add a type here? ```with types; listOf packages```?

File: nixos/modules/installer/cd-dvd/sd-image.nix [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+        Derivations to be included in the Nix store in the generated SD image.
+      '';
+    };
+
+    bootSize = mkOption {
+      default = 128;
+      description = ''
+        Size of the /boot partition.
```
Comment (bjornfor) [inline nixos/modules/installer/cd-dvd/sd-image.nix]: Missing type. And could you add the unit (KiB/MiB...?) in the description?

**Comment by bjornfor** [comment]
Great stuff!

Question. I know that Buildroot uses genext2fs to generate ext2 filesystems (which are easily transformed into ext3/4). Looking at the code you wrote using debugfs from e2fsprogs, it looks like a (relatively) cumbersome tool. Did you consider using genext2fs?

**Comment by dezgeg** [comment]
IIRC when I tried `genext2fs` from nixpkgs it was unusably slow for large images (presumably `O(n^2)` algorithm somewhere)

**Comment by bjornfor** [comment]
@dezgeg: Interesting (re. slow genext2fs).

**Comment by dezgeg** [comment]
I squashed in this diff:
````
diff --git a/nixos/modules/installer/cd-dvd/sd-image.nix b/nixos/modules/installer/cd-dvd/sd-image.nix
index 0356756..12b4f30 100644
--- a/nixos/modules/installer/cd-dvd/sd-image.nix
+++ b/nixos/modules/installer/cd-dvd/sd-image.nix
@@ -21,6 +21,7 @@ in
 {
   options.sdImage = {
     storePaths = mkOption {
+      type = with types; listOf package;
       example = literalExample "[ pkgs.stdenv ]";
       description = ''
         Derivations to be included in the Nix store in the generated SD image.
Stage this hunk [y,n,q,a,d,/,j,J,g,e,?]? y
@@ -28,9 +29,10 @@ in
     };
 
     bootSize = mkOption {
+      type = types.int;
       default = 128;
       description = ''
-        Size of the /boot partition.
+        Size of the /boot partition, in megabytes.
       '';
     };
````

**Comment by bjornfor** [comment]
Hm, I thought the sd-card derivations were meant to be evaluated on x86 machine (cross-compile), but  doesn't look like it(?), seeing as the name "sd-image-x86_64-linux.img" appears in the output and that the build fails:

```
$ nix-build nixos -A config.system.build.sdImage -I nixos-config=nixos/modules/installer/cd-dvd/sd-image-armv7l-multiplatform.nix
warning: dumping very large path (> 256 MiB); this may run out of memory
download-from-binary-cache.pl: still waiting for ‘https://cache.nixos.org/2j2hbvvs9l58rs1ml6v5k43z18rv7y2n.narinfo’ after 5 seconds...
these derivations will be built:
  /nix/store/dysh0m5qzvrqhx65an2ib0vxmwc9ga1l-nixos-15.07.git.c8ee2af.drv
  /nix/store/4djnpmg8qwbq1niizgr7f0rjd533na6n-local-cmds.drv
  /nix/store/a0d6im0nfwcbfba4gnlb9qxvad1vbqcg-stage-2-init.sh.drv
  /nix/store/93j8bhv9xh8pg20xg8a5lrx21kikg2w9-nixos-15.07.git.c8ee2af.drv
  /nix/store/891s6103zf195ilbzp2181k4qdd9abmm-ext4-fs.img.drv
  /nix/store/6az0hin4yh9sx0w8798w2kw4a8jdxwv0-sd-image-x86_64-linux.img.drv
building path(s) ‘/nix/store/7mixz7bh33znig4b4lnl8wpbpq5vmzvg-nixos-15.07.git.c8ee2af’
building path(s) ‘/nix/store/j7fgqq1d8jpr6gg38y8sp72xn755p08y-local-cmds’
building path(s) ‘/nix/store/jjscf27319h7splgjgy1fpg73a36mghr-stage-2-init.sh’
building path(s) ‘/nix/store/l34xh4z39swji7cx8pgsg2mz9ndxmlij-nixos-15.07.git.c8ee2af’
building path(s) ‘/nix/store/ph884s122jcqknbp1ygq58hrkdlzygwh-ext4-fs.img’
Creating an EXT4 image of 2060423168 bytes (numInodes=94634, numDataBlocks=313765)
mke2fs 1.42.13 (17-May-2015)
ext2fs_check_if_mount: Can't check if filesystem is mounted due to missing mtab file while determining whether /nix/store/ph884s122jcqknbp1ygq58hrkdlzygwh-ext4-fs.img is mounted.
Discarding device blocks: done                            
Creating filesystem with 503033 4k blocks and 125952 inodes
Filesystem UUID: 44444444-4444-4444-8888-888888888888
Superblock backups stored on blocks: 
	32768, 98304, 163840, 229376, 294912

Allocating group tables: done                            
Writing inode tables: done                            
Creating journal (8192 blocks): done
Writing superblocks and filesystem accounting information: done 

/nix/store/p75sm846yb0pdxziiy6982sknjynsdn9-stdenv/setup: line 846: 040000 | 0/nix/store/7mixz7bh33znig4b4lnl8wpbpq5vmzvg-nixos-15.07.git.c8ee2af/nixos/nixpkgs/popcorntime-0.3.7.2-src/node_modules/adm-zip/test/assets/attributes_test 555: division by 0 (error token is "nix/store/7mixz7bh33znig4b4lnl8wpbpq5vmzvg-nixos-15.07.git.c8ee2af/nixos/nixpkgs/popcorntime-0.3.7.2-src/node_modules/adm-zip/test/assets/attributes_test 555")
builder for ‘/nix/store/891s6103zf195ilbzp2181k4qdd9abmm-ext4-fs.img.drv’ failed with exit code 1
cannot build derivation ‘/nix/store/6az0hin4yh9sx0w8798w2kw4a8jdxwv0-sd-image-x86_64-linux.img.drv’: 1 dependencies couldn't be built
error: build of ‘/nix/store/6az0hin4yh9sx0w8798w2kw4a8jdxwv0-sd-image-x86_64-linux.img.drv’ failed
```

Wait... the error is in "nix/store/7mixz7bh33znig4b4lnl8wpbpq5vmzvg-nixos-15.07.git.c8ee2af/nixos/nixpkgs/popcorntime-0.3.7.2-src/....", why does it pull in that?! I have popcorntime-0.3.7.2-src/ as an "untracked" directory in my git working tree. It shouldn't be referenced by anything.

After removing popcorntime-0.3.7.2-src/ and restarting build, it works (build completes). But it prints this suspicious warning:

```
$ nix-build nixos -A config.system.build.sdImage -I nixos-config=nixos/modules/installer/cd-dvd/sd-image-armv7l-multiplatform.nix
...
mkfs.fat 3.0.28 (2015-05-16)
/nix/store/hclz1k77i74av9ph4ymwsklj5bhkp3ww-extlinux-conf-builder.sh: line 121: cd: /nix/var/nix/profiles: No such file or directory
```

**Comment by dezgeg** [comment]
Nope, these images can't be cross compiled (maybe it should indeed assert that).

As for the untracked popcorn-time thing, https://github.com/NixOS/nixpkgs/blob/6948c3ab80797770bc584e521b5c5a8ca02b495e/nixos/modules/profiles/clone-config.nix probably pulls it (I don't know exactly how, but... ;)

**Comment by dezgeg** [comment]
I squashed in this and added a fix for the `/nix/store/hclz1k77i74av9ph4ymwsklj5bhkp3ww-extlinux-conf-builder.sh: line 121: cd: /nix/var/nix/profiles: No such file or directory` issue.

````
diff --git nixos/modules/installer/cd-dvd/sd-image-armv7l-multiplatform.nix nixos/modules/installer/cd-dvd/sd-image-armv7l-multiplatform.nix
index 12b4c84..0ca57a4 100644
--- nixos/modules/installer/cd-dvd/sd-image-armv7l-multiplatform.nix
+++ nixos/modules/installer/cd-dvd/sd-image-armv7l-multiplatform.nix
@@ -13,10 +13,17 @@ in
     ./sd-image.nix
   ];
 
+  assertions = lib.singleton {
+    assertion = pkgs.stdenv.system == "armv7l-linux";
+    message = "sd-image-armv7l-multiplatform.nix can be only built natively on ARMv7; " +
+      "it cannot be cross compiled";
+  };
+
   boot.loader.grub.enable = false;
   boot.loader.generic-extlinux-compatible.enable = true;
 
-  boot.kernelPackages = pkgs.linuxPackages_testing; # FIXME: change this to linuxPackages_latest once v4.2 is out
+  # FIXME: change this to linuxPackages_latest once v4.2 is out
+  boot.kernelPackages = pkgs.linuxPackages_testing;
   boot.kernelParams = ["console=ttyS0,115200n8" "console=ttyAMA0,115200n8" "console=tty0"];
 
   # FIXME: fix manual evaluation on ARM
diff --git nixos/modules/installer/cd-dvd/sd-image-raspberrypi.nix nixos/modules/installer/cd-dvd/sd-image-raspberrypi.nix
index 35cc8be..199a252 100644
--- nixos/modules/installer/cd-dvd/sd-image-raspberrypi.nix
+++ nixos/modules/installer/cd-dvd/sd-image-raspberrypi.nix
@@ -13,6 +13,12 @@ in
     ./sd-image.nix
   ];
 
+  assertions = lib.singleton {
+    assertion = pkgs.stdenv.system == "armv6l-linux";
+    message = "sd-image-raspberrypi.nix can be only built natively on ARMv6; " +
+      "it cannot be cross compiled";
+  };
+
   # Needed by RPi firmware
   nixpkgs.config.allowUnfree = true;

````

**Comment by wmertens** [comment]
Just dropping in to say :heart_eyes:

**Comment by dezgeg** [comment]
So all the issues and improvements here should have been noted, I suppose this could be merged. I see I've forgotten to ping @viric here.

**Comment by viric** [comment]
I don't think it will break anything from anyone, so let's have this in. Thank you for your work! Definitely, the cc reminded me of this merge.

**Comment by dvc94ch** [comment]
@dezgeg the latest arm6 image from https://www.cs.helsinki.fi/u/tmtynkky/nixos-arm/installer/ doesn't work for me. Running `curl` causes `Illegal instruction (core dumped)` on a RPI1. This causes `nixos-channel --update` to fail which I think causes `nixos-rebuild` to not find `nixpkgs/nixos` in the nix search path.

**Pull request merged**